packages/dds/tree/src/simple-tree/api/schemaFactory.ts

```diff
@@ -472,15 +501,20 @@ export class SchemaFactory<
 	 *
 	 * {@label STRUCTURAL}
 	 */
-	public array<const T extends TreeNodeSchema | readonly TreeNodeSchema[]>(
+	public array<
+		const T extends TreeNodeSchema | readonly TreeNodeSchema[],
+		const TCustomMetadata = unknown,
```

Comment:
I don't think adding "props" to the structurally typed APIs are a good idea.

If two people call this with different props but the same types, will they get back the same schema, discarding the props who ever was second gave, or get back differed schema with colliding names? Or do we expect the props to be somehow converted into strings and included in the structurally generated name? Given the type constraint is just "unknown" that does not seem viable.

I see no possible way this API could produce reasonable results, and its undocumented (and I assume untested since don't think there is a correct thing the implementation could do here)

Reply:
That's an interesting point. I admit that I haven't thought about the case where you define a schema twice with the same name. I assume that our _current_ behavior (and guarantee?) is that you get back the "same thing" in both cases? In which case, yea - I agree this is weird that you can do it twice with difference metadata. I can't think of anything to do in that case except throw.

Reply:
`props` isn't intended to influence the schema identity and should only serve as a supplementary parameter to the schema factory right? Is there no way for the schema factory to disregard the `props` parameter when creating the schema instance?

Reply:
> `props` isn't intended to influence the schema identity and should only serve as a supplementary parameter to the schema factory right? Is there no way for the schema factory to disregard the `props` parameter when creating the schema instance?

The data inside of the `props` parameter can end up in the actual schema objects. The `metadata` property added here is such an example. So this is a problem.

It's probably fine to only support additional options for named schemas. We can follow up on the structurally typed side of things in the future as needed. I'll take a stab at refactoring this soon.

Reply:
What do you think we should do if the user does `sf.object("Foo", { metadata: X })` and then later does `sf.object("Foo", { metadata: Y })`? We could:

1. Fail unconditionally if you try to make a schema with the same name twice
2. Fail if you try to make a schema with the same name twice _and_ either creation includes metadata
3. Allow the first creation to establish metadata, but no others
4. If there's metadata, diff it (how?), and fail if it's different
5. Something else?

Reply:
Sorry, was heads down on the website update for the last month+. Getting back to this now.

On the surface, I like \#1 above best. _But_, I don't think we can reasonably depend on tree shaking, etc. to ensure that the same schema definition code doesn't get run multiple times in an application, so presumably we need to be more tolerant.

For the same reason as above, I think we ultimately _have_ to do something like what you suggested in \#4, if we want to expose the metadata the way I prototyped here.

Reply:
That said, since this is only true (correct me if I am wrong) for structurally typed schemas, maybe we can get away with only supporting this (at least to start) for class-based schema?

This case seems simpler, and generally we (again, correct me if I'm wrong) would prefer for users to leverage class-based schemas for most scenarios. We could always come back to this (i.e., support node metadata for structurally based schemas) if/when there is a user need.

Reply:
FWIW I've since checked in `areMetadataEqual` - might be what we want for the comparison